File: sdk/python/memg/conversation.py
"""Conversation message normalization and deduplication."""

from __future__ import annotations

from typing import List


def normalize_conversation_messages(messages: List[dict]) -> List[dict]:
    """Filter to user/assistant roles, trim content, remove empty messages.

    Logic matches Go memory/conversation.go NormalizeConversationMessages.
    """
    out = []
    for msg in messages:
        if msg is None:
            continue
        role = (msg.get("role") or "").strip()
        if role not in ("user", "assistant"):
            continue
        content = (msg.get("content") or "").strip()
        if not content:
            continue
        out.append({"role": role, "content": content})
    return out


def _overlap_length(existing: List[dict], incoming: List[dict]) -> int:
    """Find the longest tail of existing that matches a prefix of incoming."""
    existing_len = len(existing)
    max_overlap = min(len(incoming), existing_len)
    for overlap in range(max_overlap, 0, -1):
        match = True
        for i in range(overlap):
            e = existing[existing_len - overlap + i]
            m = incoming[i]
            if e.get("role") != m.get("role") or e.get("content") != m.get("content"):
                match = False
                break
        if match:
            return overlap
    return 0


def diff_incoming_messages(existing: List[dict], incoming: List[dict]) -> List[dict]:
    """Return only messages from incoming that are not already in existing.

    Logic matches Go memory/conversation.go DiffIncomingMessages.
    Uses sequence-based tail comparison.
    """
    incoming = normalize_conversation_messages(incoming)
    existing = normalize_conversation_messages(existing)

    if not existing:
        return incoming

    overlap = _overlap_length(existing, incoming)
    if overlap >= len(incoming):
        return []
    return incoming[overlap:]


def merge_history(history: List[dict], incoming: List[dict]) -> List[dict]:
    """Merge stored history with incoming messages, deduplicating overlap.

    Logic matches Go memory/conversation.go MergeHistory.
    """
    history = normalize_conversation_messages(history)
    incoming = normalize_conversation_messages(incoming)

    if not history:
        return incoming
    if not incoming:
        return history

    overlap = _overlap_length(history, incoming)
    merged = list(history)
    merged.extend(incoming[overlap:])
    return merged

File: sdk/typescript/src/conversation.ts
/**
 * Conversation message utilities: normalize, diff, and merge.
 * Matches the Go memory/conversation.go logic.
 */

interface Message {
  role: string;
  content: string;
}

/**
 * Drop non-conversation roles (e.g., system prompts), trim whitespace,
 * and return only user/assistant turns with non-empty content.
 */
export function normalizeConversationMessages(
  messages: Message[]
): Message[] {
  const out: Message[] = [];
  for (const msg of messages) {
    if (!msg) continue;
    const role = msg.role.trim();
    if (role !== 'user' && role !== 'assistant') continue;
    const content = msg.content.trim();
    if (content === '') continue;
    out.push({ role, content });
  }
  return out;
}

/**
 * Return only messages from `incoming` that are not already present in
 * `existing`. Uses sequence-based tail comparison (the tail of existing
 * matching the prefix of incoming) so that legitimately repeated messages
 * are not filtered out.
 */
export function diffIncomingMessages(
  existing: Message[],
  incoming: Message[]
): Message[] {
  const normalizedExisting = normalizeConversationMessages(existing);
  const normalizedIncoming = normalizeConversationMessages(incoming);

  if (normalizedExisting.length === 0) return normalizedIncoming;

  const overlap = overlapLength(normalizedExisting, normalizedIncoming);

  if (overlap >= normalizedIncoming.length) return [];
  return normalizedIncoming.slice(overlap);
}

/**
 * Merge stored history with the latest incoming request without duplicating
 * turns the client already replayed. Returns history + incoming[overlap:].
 */
export function mergeHistory(
  history: Message[],
  incoming: Message[]
): Message[] {
  const normalizedHistory = normalizeConversationMessages(history);
  const normalizedIncoming = normalizeConversationMessages(incoming);

  if (normalizedHistory.length === 0) return normalizedIncoming;
  if (normalizedIncoming.length === 0) return normalizedHistory;

  const overlap = overlapLength(normalizedHistory, normalizedIncoming);
  const merged: Message[] = [];
  merged.push(...normalizedHistory);
  merged.push(...normalizedIncoming.slice(overlap));
  return merged;
}

/**
 * Find the longest overlap between the tail of `existing` and the prefix
 * of `incoming`.
 */
function overlapLength(existing: Message[], incoming: Message[]): number {
  const existingLen = existing.length;
  let maxOverlap = incoming.length;
  if (maxOverlap > existingLen) maxOverlap = existingLen;

  for (let overlap = maxOverlap; overlap > 0; overlap--) {
    let match = true;
    for (let i = 0; i < overlap; i++) {
      const e = existing[existingLen - overlap + i];
      const m = incoming[i];
      if (e.role !== m.role || e.content !== m.content) {
        match = false;
        break;
      }
    }
    if (match) return overlap;
  }
  return 0;
}
